crates/miden-objects/src/note/note_tag.rs

```diff
@@ -248,23 +276,48 @@ impl fmt::Display for NoteTag {
 
 impl From<u32> for NoteTag {
     fn from(value: u32) -> Self {
-        Self(value)
+        // Mask out the two most significant bits.
+        match value & 0xc0000000 {
+            NETWORK_ACCOUNT => {
+                // Network, Specific Account (lower 30 bits are the account ID)
+                Self::NetworkAccount(value)
+            },
+            NETWORK_USECASE => {
+                // NetworkUse case (split lower 30 bits into two u16)
+                Self::NetworkUseCase((value >> 16) as u16, value as u16)
+            },
+            LOCAL_ACCOUNT => {
+                // Local, Any (lower 30 bits are the account ID)
+                Self::LocalAccount(value)
+            },
+            LOCAL_USECASE => {
+                // Local, Any (split lower 30 bits into two u16)
+                Self::LocalUseCase((value >> 16) as u16, value as u16)
+            },
+            _ => {
+                // This branch should be unreachable because `prefix` is derived from
+                // the top 2 bits of a u32, which can only be 0, 1, 2, or 3.
+                unreachable!("Invalid value encountered: {:b}", value);
+            },
+        }
     }
 }
 
 impl TryFrom<u64> for NoteTag {
     type Error = TryFromIntError;
 
     fn try_from(value: u64) -> Result<Self, Self::Error> {
-        Ok(Self(value.try_into()?))
+        let value_u32: u32 = value.try_into()?;
+        Ok(NoteTag::from(value_u32))
     }
 }
 
 impl TryFrom<Felt> for NoteTag {
     type Error = TryFromIntError;
 
     fn try_from(value: Felt) -> Result<Self, Self::Error> {
-        Ok(Self(value.as_int().try_into()?))
+        let int_u32: u32 = value.as_int().try_into()?;
+        Ok(NoteTag::from(int_u32))
     }
 }
```

Comment:
I think I would remove these, they don't seem very useful. We also don't use these anywhere in miden-base, so chances are they are not used at all. If we have to convert a `Felt` to a `NoteTag` somewhere, then we should handle the conversion to u32 in that code, imo.

If we do end up needing them, we can always bring them back easily.